### ISSUE TITLE
False Positive in FallThroughCheck

### ISSUE THREAD

**Issue by Kevin222004** [checkstyle/checkstyle#13545]
I have read check documentation: https://checkstyle.org/checks/coding/fallthrough.html#FallThrough
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

-------

**Config** 

```
kevin@inspiron-15-5510:~/Desktop/check_style$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="FallThrough">
            <property name="checkLastCaseGroup" value="true"/>
        </module>
    </module>
</module>
```

**Code**

```
kevin@inspiron-15-5510:~/Desktop/check_style$ cat Test.java
     1  public class Test {
     2      void method(int i) {
     3        switch(i) {
     4          case 0: 
     5            int j = 22; // fallthru  
     6          case 1: /* line no :- 6 */
     7            int k = 2;
     8            break;
     9          case 2:
    10            i++;
    11            // fall thru  
    12          case 3: 
    13            i++;
    14            break;  
    15        }   
    16      }
    17      
    18      void foo(int i){
    19        switch(i) {
    20          case 1:
    21            int fallthru = 2; // fallthru
    22          case 2:   
    23            int fallthrough = 2; //            fall thru
    24          case 3: 
    25            break;
    26          case 4:
    27            int fallthru2 = 3;
    28            //     fall   -   thru
    29          case 5: 
    30            int fallthru3 = 4; // This is a fallthru comment
    31          case 6: 
    32            int fallthru4 = 5; // falldown
    33            default:
    34          }
    35      }
    36    }
```

**Output**

```
kevin@inspiron-15-5510:~/Desktop/check_style$ java \
   -jar /home/kevin/Downloads/checkstyle-10.12.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:22:7: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] Test.java:24:7: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] Test.java:29:7: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] Test.java:31:7: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] Test.java:33:9: Fall through from previous branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 5 errors.
```

-----------

**Expectation**
If we are showing violation on line 22 and line 24 Then line no 6 also should show violation

**Comment by Kevin222004** [comment]
Just an Opinion 

In https://github.com/checkstyle/checkstyle/pull/12966 I have generated lots of reports regarding this check and see lots of project code in the report. The majority of the projects in which the switch contains fall-thru have mentioned the reason related to fall-thru and for that reason multiple times this keyword fall-thru is mentioned. and most cases this kind of comment is related to the reason why the case is fall-thru it is after the logic of that case. And if they are mentioning this case they might be aware that this is fall-thru and it's intentional.

My view is why are we supposed to https://checkstyle.org/checks/coding/fallthrough.html#FallThrough:~:text=and%20must%20be%20on%20the%20last%20non%2Dempty%20line%20before%20the%20case%20triggering%20the%20warning%20or%20on%20the%20same%20line%20before%20the%20case%20(ugly%2C%20but%20possible) have this kind of rule that force to write a new comment `fall-thru` on new last empty line even they mentioned their reason above it and in that reason they multiple time use same-keyword. Isn't that annoying to them?  

we should only look for the comment `fall-thru` from newline in a case group when the logic of that case is completed.

**Comment by romani** [comment]
I confirming that there is no violation on line 6, that is weird.

@rnveach , please approve issue if you agree.

**Comment by nrmancuso** [comment]
> My view is why are we supposed to https://checkstyle.org/checks/coding/fallthrough.html#FallThrough:~:text=and%20must%20be%20on%20the%20last%20non%2Dempty%20line%20before%20the%20case%20triggering%20the%20warning%20or%20on%20the%20same%20line%20before%20the%20case%20(ugly%2C%20but%20possible) have this kind of rule that force to write a new comment fall-thru on new last empty line

It is likely this was a restriction based on the check using `getFileContents` and should probably be lifted now that we fully analyze the AST in this check.

**Comment by Kevin222004** [comment]
@rnveach ping

Please approve if you agree

**Comment by nrmancuso** [comment]
@Kevin222004 please do `cat -n` if you are referencing line numbers

**Comment by Kevin222004** [comment]
```
     1  public class Test {
     2      void method(int i) {
     3        switch(i) {
     4          case 0: 
     5            int j = 22; // fallthru  
     6          case 1: /* line no :- 6 */
     7            int k = 2;
     8            break;
     9          case 2:
    10            i++;
    11            // fall thru  
    12          case 3: 
    13            i++;
    14            break;  
    15        }   
    16      }
    17      
    18      void foo(int i){
    19        switch(i) {
    20          case 1:
    21            int fallthru = 2; // fallthru
    22          case 2:   
    23            int fallthrough = 2; //            fall thru
    24          case 3: 
    25            break;
    26          case 4:
    27            int fallthru2 = 3;
    28            //     fall   -   thru
    29          case 5: 
    30            int fallthru3 = 4; // This is a fallthru comment
    31          case 6: 
    32            int fallthru4 = 5; // falldown
    33            default:
    34          }
    35      }
    36    }
```

**Comment by Kevin222004** [comment]
@nrmancuso I updated the description

**Comment by nrmancuso** [comment]
Ok, IMO, we need to consider ANY comment containing the magic words in the previous case to allow fall through with no violation. The following should all be ok:

```
        switch (i) {
            case 1:
                // fall through
                System.out.println("1");
            case 2:
                System.out.println("2");
                // fall through
            case 3:
                System.out.println("3");
                /* fall through */
            case 4:
                System.out.println("4");// fall through
            case 5: // fall through
                System.out.println("5");
            case 6:
                System.out.println("6"); /* fall through */
            case 7:
                System.out.println("7");
                /* fall
                    * through */
            default:
                System.out.println("default");
        }



```

My thought is this: if someone is intentionally writing the words "fall through" in a switch case statement, they are aware of the semantics of what they wrote and intend for cases to fall through. Maybe we could add a strict property that keeps the old behavior? Or the other way around to not break existing usages.

**Comment by rnveach** [comment]
> If we are showing violation on line 22 and line 24 Then line no 6 also should show violation
>     21            int fallthru = 2; // fallthru
    22          case 2:   
    23            int fallthrough = 2; //            fall thru
    24          case 3: 

Why is line 22 and 24 a violation?

https://checkstyle.org/checks/coding/fallthrough.html#Description
>  The comment containing these words must be all on one line, and must be on the last non-empty line before the case triggering the warning or on the same line before the case 
> default value: "falls?[ -]?thr(u|ough)"

We clearly have the matching suppression text. Suppression comment is all on 1 line. Suppression comment is on the previous line before the violation case.

It seems to me that violations on line 22 and 24 are wrong, and they are the false positive. Not line 6.

**Comment by rnveach** [comment]
@nrmancuso Your case 1 and 5 should be a violation. Suppression comment is not last line before violation case. Changing this would have to be another issue, and I truthfully don't agree with it.

Case 7 would also be a violation and would have to be another issue. While I could get behind support for multi-line comment suppression, I question accurately supporting finding/removing a random `*` starting a new line.

**Comment by Kevin222004** [comment]
@rnveach 
> It seems to me that violations on line 22 and 24 are wrong, and they are the false positive. Not line 6. 

then we can close this issue because from 10.2.4 
for same code and config in description 
```
kevin@kevin-Inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Downloads/checkstyle-10.12.4-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/kevin/Desktop/check_style/Test.java:29:9: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] /home/kevin/Desktop/check_style/Test.java:33:11: Fall through from previous branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 2 errors.
```

This issue is may be opened to point out this I will find the exact discussion and put it out here.

**Comment by rnveach** [comment]
@romani Please confirm if you agree to close this as already fixed.

**Comment by romani** [comment]
confirming , behavior shown at https://github.com/checkstyle/checkstyle/issues/13545#issuecomment-1752148444 is same on latest release 10.12.4, and 2 violations is expected.

So this issue is resolved by some other update.
